components/keyrings/keyring_kms/backend/kms.h

```diff
@@ -0,0 +1,403 @@
+
+#ifndef __KMS_AWS_H__
+#define __KMS_AWS_H__
+
+// Code based on xbcloud
+
+#include <curl/curl.h>
+#include <algorithm>
+#include <cstring>
+#include <functional>
+#include <map>
+#include <memory>
+#include <queue>
+#include <string>
+#include <vector>
+
+#include <openssl/hmac.h>
+#include <openssl/sha.h>
+
+namespace aws {
+
+std::string uri_escape_string(const std::string &s);
+std::string uri_escape_path(const std::string &path);
+bool http_init();
+void http_cleanup();
+
+using curl_easy_unique_ptr =
+    std::unique_ptr<::CURL, decltype(&curl_easy_cleanup)>;
+inline curl_easy_unique_ptr make_curl_easy() {
+  return curl_easy_unique_ptr(curl_easy_init(), curl_easy_cleanup);
+}
+
+template <typename T>
+std::vector<unsigned char> sha256(const T &s) {
+  std::vector<unsigned char> md(SHA256_DIGEST_LENGTH);
+
+  SHA256_CTX sha256;
+  SHA256_Init(&sha256);
+  SHA256_Update(&sha256, &s[0], s.size());
```

Comment:
```suggestion
  SHA256_Update(&sha256, s.c_str(), s.size());
```
`c_str()` is safer here

Reply:
That won't work, it's not a string, but an instance of Http_buffer